IdentityCore/src/throttling/cache/MSIDLRUCache.m

```diff
@@ -0,0 +1,432 @@
+                               error:(NSError **)error
+{
+    if (![self.container objectForKey:signature])
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"MSIDLRUCache Error: Unable to find valid node for the input signature during retrieval", nil, nil, nil, nil, nil, YES);
+        }
+        return nil;
+    }
+    
+    //retrieve node
+    MSIDLRUCacheNode *node = [self.container objectForKey:signature];
+    
+    //remove from current cache slot
+    [self removeFromCacheImpl:signature
+                        error:error];
+    //move to front
+    [self addToFrontImpl:node];
+    
+    return node.cacheRecord;
+}
+
+- (void)addToFront:(MSIDLRUCacheNode *)node
+{
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        [self addToFrontImpl:node];
+    });
+}
+
+- (void)addToFrontImpl:(MSIDLRUCacheNode *)node
+{
+    NSString *currentHeadSignature = self.head.nextSignature; //node currently pointed by the head
+    MSIDLRUCacheNode *currentHeadNode = [self.container objectForKey:currentHeadSignature];
+    
+    /**
+    BEFORE:
+     A->B->C
+     A<-B<-C
+    AFTER:
+     A->C
+     A<-C
+     */
+    currentHeadNode.prevSignature = [node.signature mutableCopy];
+    node.prevSignature = [HEAD_SIGNATURE mutableCopy];
+    node.nextSignature = [currentHeadSignature mutableCopy];
+    self.head.nextSignature = [node.signature mutableCopy];
+    
+    [self.container setObject:currentHeadNode forKey:currentHeadSignature];
+    [self.container setObject:node forKey:node.signature];
+}
+
+- (MSIDLRUCacheNode *)getHeadNode
+{
+    __block MSIDLRUCacheNode *node;
+    dispatch_sync(self.synchronizationQueue, ^{
+        if (![self.head.nextSignature isEqualToString:TAIL_SIGNATURE])
+        {
+            node = [self.container objectForKey:self.head.nextSignature];
+        }
+    });
+    return node;
+    
+}
+
+- (MSIDLRUCacheNode *)getTailNode
+{
+    __block MSIDLRUCacheNode *node;
+    dispatch_sync(self.synchronizationQueue, ^{
+        if (![self.tail.prevSignature isEqualToString:HEAD_SIGNATURE])
+        {
+            node = [self.container objectForKey:self.tail.prevSignature];
+        }
+    });
+    return node;
+}
+
+- (NSArray *)enumerateAndReturnAllObjects
+{
+    __block NSMutableArray *res;
+    dispatch_sync(self.synchronizationQueue, ^{
+        res = [self enumerateAndReturnAllObjectsImpl];
+    });
+    return res;
+}
+
+- (NSMutableArray *)enumerateAndReturnAllObjectsImpl
+{
+    NSMutableArray *res;
+    if (![self.head.nextSignature isEqualToString:TAIL_SIGNATURE])
+    {
+        res = [NSMutableArray new];
+        NSMutableString *signature = [self.head.nextSignature mutableCopy];
+        
+        while (![signature isEqualToString:TAIL_SIGNATURE])
+        {
+            MSIDLRUCacheNode *node = [self.container objectForKey:signature];
+            [signature setString:node.nextSignature];
+            [res addObject:node.cacheRecord];
```

Comment:
will a nil check here prevent us from crash? :)

Reply:
good catch. I should remove nullable flag from addToCache param, by doing so it will enforce non-null input into addToCache param, and by doing so, enumerate API will also ensure no such case arises.

Reply:
> by doing so it will enforce non-null input

It is still possible to pass null even if param is declared as "no-null". It safe to check for nil before you add to the array.